src/wayland/input_method/input_method_popup_surface.rs

```diff
@@ -0,0 +1,87 @@
+use std::sync::{Arc, Mutex};
+
+use wayland_protocols_misc::zwp_input_method_v2::server::zwp_input_popup_surface_v2::{
+    self, ZwpInputPopupSurfaceV2,
+};
+use wayland_server::backend::{ClientId, ObjectId};
+use wayland_server::{protocol::wl_surface::WlSurface, Dispatch};
+
+use crate::utils::{Logical, Point};
+
+use super::InputMethodManagerState;
+
+#[derive(Default, Debug)]
+pub(crate) struct InputMethodPopupSurface {
+    pub surface_role: Option<ZwpInputPopupSurfaceV2>,
+    pub surface: Option<WlSurface>,
+    x: i32,
+    y: i32,
+    width: i32,
+    height: i32,
+    l_x: i32,
+    l_y: i32,
+}
+
+/// Handle to an input method instance
+#[derive(Default, Debug, Clone)]
+pub struct InputMethodPopupSurfaceHandle {
+    pub(crate) inner: Arc<Mutex<InputMethodPopupSurface>>,
+}
+
+impl InputMethodPopupSurfaceHandle {
+    /// Used to store surface coordinates
+    pub fn add_coordinates(&self, x: i32, y: i32, width: i32, height: i32) {
+        let mut inner = self.inner.lock().unwrap();
+        inner.x = x;
+        inner.y = y;
+        inner.width = width;
+        inner.height = height;
+    }
+
+    /// Used to access the relative location of an input popup surface
+    pub fn coordinates(&self) -> (i32, i32, i32, i32) {
+        let inner = self.inner.lock().unwrap();
+        (
+            inner.x + inner.l_x,
+            inner.y + inner.l_y,
+            inner.width,
+            inner.height,
+        )
+    }
```

Comment:
Have you considered using Smithay's Rectangle type, instead of tupe? It is usually used to store position + size.

Reply:
Good suggestion, I will update

Reply:
Updated